### PR TITLE
Add min version for Catalyst to `qml.compiler`

### PR DESCRIPTION
**Pull request by maliasadi**
The new PL version should check that the Catalyst version, if installed, is at least `v0.4.0`.

Note there are a couple of TODOs to address after releasing Catalyst `v0.4.0`.

[sc-52634]

### CODE REVIEW

File: pennylane/compiler/compiler.py [hunk 1]
```diff
@@ -18,6 +18,11 @@
 from importlib import reload, metadata
 from collections import defaultdict
 import dataclasses
+import os
+
+from packaging import version
+
+MIN_CATALYST_VERSION = version.parse(os.environ.get("MIN_CATALYST_VERSION", "0.4.0"))
```
Comment (timmysilv) [inline pennylane/compiler/compiler.py:25]:
```suggestion
from semantic_version import Version

MIN_CATALYST_VERSION = Version(os.environ.get("MIN_CATALYST_VERSION", "0.4.0"))
```

File: pennylane/compiler/compiler.py [hunk 2]
```diff
@@ -65,6 +70,14 @@ def _refresh_compilers():
         module_name = entry.module.split(".")[0]
         AvailableCompilers.names_entrypoints[module_name][entry.name] = entry
 
+    # Check the minimum version of Catalyst if installed
+    if "catalyst" in AvailableCompilers.names_entrypoints:
+        catalyst_version = version.parse(metadata.version("pennylane-catalyst"))
```
Comment (timmysilv) [inline pennylane/compiler/compiler.py:75]:
```suggestion
        catalyst_version = Version(metadata.version("pennylane-catalyst"))
```

File: setup.py [hunk 3]
```diff
@@ -33,6 +33,7 @@
     "pennylane-lightning>=0.33",
     "requests",
     "typing_extensions",
+    "packaging",
```
Comment (timmysilv) [inline setup.py:36]:
```suggestion
```

File: .github/workflows/interface-unit-tests.yml [hunk 4]
```diff
@@ -166,7 +166,8 @@ jobs:
       install_pennylane_lightning_master: true
       pytest_coverage_flags: ${{ inputs.pytest_coverage_flags }}
       pytest_markers: torch and not qcut and not finite-diff and not param-shift
-
+      env: # TODO: remove this after Catalyst v0.4.0 release
+        MIN_CATALYST_VERSION: "0.3.2"
```
Comment (timmysilv) [inline .github/workflows/interface-unit-tests.yml:170]: I feel weird adding the handling of this environment variable to source code if it's only for this test suite. Is that the case here? Also, is this intended to only be in `torch-tests`?

**Review by timmysilv — COMMENTED** [review]
made some suggestions to use `semantic-version` to avoid adding another dependency to PennyLane (not sure if I got the syntax exactly right, but I think it should do)

File: pennylane/compiler/compiler.py (hunk 1 above)
Comment (josh146) [reply pennylane/compiler/compiler.py:25]: thanks @timmysilv :+1:

Comment (maliasadi) [reply pennylane/compiler/compiler.py:25]: nice, I discussed this with Lee in advanced about common practices here. I wasn't aware of `semantic_version` :+1:

File: .github/workflows/interface-unit-tests.yml (hunk 4 above)
Comment (maliasadi) [reply .github/workflows/interface-unit-tests.yml]: This env variable is added to `external-libraries-tests` [0f32cb9](https://github.com/PennyLaneAI/pennylane/pull/4979/commits/0f32cb93db6b1f112c7db1bbc561a20b61370d6f) to temporary alter the minimum version of Catalyst for tests in `test_compiler.py`. This will be removed after Catalyst v0.4.0 release next week.

Comment (timmysilv) [reply .github/workflows/interface-unit-tests.yml]: does this need to be in the 0.34 release then? we've already made the RC branch, so if so, please target it (`v0.34.0-rc0`) and add this PR to the 0.34 milestone to make sure we don't miss it.

Either way, is the plan after the Catalyst 0.4.0 release to hard-code the minimum to be 0.4.0?

Comment (maliasadi) [reply .github/workflows/interface-unit-tests.yml]: Thanks @timmysilv! I'll do rebase the PR. The minimum version of Catalyst is set to 0.4.0 in `compiler.py`, but I added this env variable to enable us to check the PL-Catalyst tests in `test_compiler.py` before releasing Catalyst v0.4.0. We can get rid of this env variable next week after the release.

**Review by timmysilv — APPROVED** [review]
🎉

**Review by lillian542 — APPROVED** [review]
Thanks @maliasadi! 🚀

File: pennylane/compiler/compiler.py [hunk 5]
```diff
@@ -65,6 +71,14 @@ def _refresh_compilers():
         module_name = entry.module.split(".")[0]
         AvailableCompilers.names_entrypoints[module_name][entry.name] = entry
 
+    # Check the minimum version of Catalyst if installed
+    if "catalyst" in AvailableCompilers.names_entrypoints:
+        catalyst_version = metadata.version("pennylane-catalyst")
+        if Version(re.sub(r"\.dev\d+", "", catalyst_version)) < PL_CATALYST_MIN_VERSION:
+            raise CompileError(
+                f"PennyLane-Catalyst {PL_CATALYST_MIN_VERSION} or greater is required, but installed {catalyst_version}"
```
Comment (dime10) [inline pennylane/compiler/compiler.py:79]: One downside I see with this approach is that the error is raised regardless of whether the user wants to use PennyLane with Catalyst. As long as Catalyst is installed, the error will be raised on PennyLane import 🤔 

Ideally, the error would only be raised if the user tries to actually use `qml.qjit`.

File: pennylane/compiler/compiler.py [hunk 6]
```diff
@@ -18,6 +18,12 @@
 from importlib import reload, metadata
 from collections import defaultdict
 import dataclasses
+import os
+import re
+
+from semantic_version import Version
+
+PL_CATALYST_MIN_VERSION = Version(os.environ.get("PL_CATALYST_MIN_VERSION", "0.4.0"))
```
Comment (dime10) [inline pennylane/compiler/compiler.py:26]: An environment variable determining a minimum package version is also a bit strange. Are there alternatives we can think of?

Maybe temporarily patching this value, or the version value of the installed Catalyst package, in the test configuration, until after the release? The post-release update would only touch the tests then, without needing to touch the package source.

File: pennylane/compiler/compiler.py (hunk 5 above)
Comment (dime10) [inline pennylane/compiler/compiler.py:79]: Are these lines covered? I don't see a new test.

Comment (maliasadi) [reply pennylane/compiler/compiler.py:79]: These are not covered as we update the min version to v0.3.2 to pass PL-Catalyst integration tests in `test_compiler.py` and as you mentioned these lines are called as soon as importing pennylane.

Comment (dime10) [reply pennylane/compiler/compiler.py:79]: I think it would still be customary to have a test that forces a lower version number and verifies the error is raised.

File: pennylane/compiler/compiler.py (hunk 6 above)
Comment (maliasadi) [reply pennylane/compiler/compiler.py:26]: @dime10 I only used this env variable to avoid adding any monkey patches to PL (tests) as this sort of patching should be prohibited in every codebase if possible. But, if you think monkey patching the version of Catalyst is something "okay" until the v0.4.0 release, I'd be happy to adjust the code.

File: pennylane/compiler/compiler.py (hunk 5 above)
Comment (maliasadi) [reply pennylane/compiler/compiler.py:79]: I didn't see this as a downside tbh as there's a two-way dependency between PL and Catalyst and this would encourage users to install the up-to-date and compatible Catalyst on their system. But, I can move the logic to `qml.qjit` if you think this is something unnecessary @dime10

Comment (lillian542) [reply pennylane/compiler/compiler.py:79]: Thanks for pointing that out @dime10. Brought this up at stand up and the consensus seems to be that it would be much better if the error was raised inside `qjit`.

File: pennylane/compiler/compiler.py (hunk 6 above)
Comment (dime10) [reply pennylane/compiler/compiler.py:26]: I don't agree monkey patching should necessarily be prohibited in every code base, in fact it is a common pattern for testing (https://docs.pytest.org/en/latest/how-to/monkeypatch.html), which is the use case we are discussing.
So personally I would favour it over an environment variable, but I'm also happy to let core decide here.

File: pennylane/compiler/compiler.py (hunk 5 above)
Comment (maliasadi) [reply pennylane/compiler/compiler.py]: Fixed [b9bca5d](https://github.com/PennyLaneAI/pennylane/pull/4979/commits/b9bca5d33d91ee587025f7843a80e65a9de04505) Thanks :t-rex:

File: pennylane/compiler/compiler.py (hunk 6 above)
Comment (maliasadi) [reply pennylane/compiler/compiler.py]: Thanks @dime10! I'll update the code. The core team is okay with this temporary monkey patch :+1:

File: pennylane/compiler/compiler.py [hunk 7]
```diff
@@ -41,17 +46,38 @@ class AvailableCompilers:
...
 
 
 def _refresh_compilers():
     """Scan installed PennyLane compiler packages to refresh the compilers
     names and entry points.
     """
-
-    # Refresh the list of compilers
-    AvailableCompilers.names_entrypoints = defaultdict(dict)
-
```
Comment (dime10) [inline pennylane/compiler/compiler.py:54]: Why are these lines removed?

File: tests/test_compiler.py [hunk 8]
```diff
@@ -32,6 +32,31 @@
...
+    @qml.qnode(dev)
+    def circuit():
+        qml.PauliX(0)
+        return qml.state()
+
+    with pytest.raises(
+        CompileError, match="PennyLane-Catalyst 0.4.0 or greater is required, but installed"
+    ):
+        qml.qjit(circuit)()
+    qml.compiler.compiler.AvailableCompilers.names_versions["catalyst"] = True
```
Comment (dime10) [inline tests/test_compiler.py:51]: This sort of set and reset is better handled via the `monkeypatch` fixture, check https://docs.pytest.org/en/latest/reference/reference.html#pytest.MonkeyPatch.setitem.

File: tests/test_compiler.py [hunk 9]
```diff
@@ -32,6 +32,31 @@
 # pylint: disable=too-few-public-methods, too-many-public-methods
 
 
+# TODO: Update this test for the code coverage after Catalyst 0.4.0 release
```
Comment (dime10) [inline tests/test_compiler.py:35]: I don't think we should design the test such a way that it needs updating in the future. We can write it now such that it artificially creates the conditions or state it is intended to test (without any variability).

Concretely this could mean patching the version of the installed Catalyst package to something low (like `0.1.0`), if that is possible with the method we use to check the package version, or set the minimum to something really high (like `999.0.0`).

File: pennylane/compiler/compiler.py [hunk 10]
```diff
@@ -41,17 +46,38 @@ class AvailableCompilers:
...
     # and their entry point loaders.
-    names_entrypoints = {}
+    names_entrypoints = defaultdict(dict)
+
+    # The map consists of supported compiler names (str) and their version compatibility (bool).
+    # This boolean indicates whether the installed version of a compiler package is greater
+    # than or equal to the minimum version.
+    # This value will be updated in `_check_compiler_version` to reduce the required
+    # version checks of installed compiler packages at runtime.
+    names_versions = defaultdict(dict)
```
Comment (dime10) [inline pennylane/compiler/compiler.py:56]: The caching is a nice feature 👍 

This is optional, but I would find a name for the dictionary such as `compiler_checked` more descriptive of what the code does (for instance a `false` Boolean does not indicate the compiler is incompatible, rather it leads to the version check being performed).

File: tests/test_compiler.py [hunk 11]
```diff
@@ -32,6 +32,31 @@
...
+        CompileError, match="PennyLane-Catalyst 0.4.0 or greater is required, but installed"
+    ):
+        qml.qjit(circuit)()
+    qml.compiler.compiler.AvailableCompilers.names_versions["catalyst"] = True
+
+
+# Update the compiler cached value for Catalyst version < 0.4.0 to `True`
+# for checking the rest of tests concurrently using pytest-xdist
+# TODO: Remove after Catalyst 0.4.0 release
+qml.compiler.compiler.AvailableCompilers.names_versions["catalyst"] = True
```
Comment (dime10) [inline tests/test_compiler.py:57]: Nice 👍

**Review by dime10 — CHANGES_REQUESTED** [review]
Thanks @maliasadi :)

File: tests/test_compiler.py (hunk 8 above)
Comment (maliasadi) [reply tests/test_compiler.py]: Fixed [65a1efc](https://github.com/PennyLaneAI/pennylane/pull/4979/commits/65a1efc6f99dd1236bc133900139803eb832e197)

File: tests/test_compiler.py (hunk 9 above)
Comment (maliasadi) [reply tests/test_compiler.py]: Good point! Fixed [65a1efc](https://github.com/PennyLaneAI/pennylane/pull/4979/commits/65a1efc6f99dd1236bc133900139803eb832e197)

File: pennylane/compiler/compiler.py (hunk 10 above)
Comment (maliasadi) [reply pennylane/compiler/compiler.py]: Updated [65a1efc](https://github.com/PennyLaneAI/pennylane/pull/4979/commits/65a1efc6f99dd1236bc133900139803eb832e197)

File: tests/test_compiler.py (hunk 9 above)
Comment (maliasadi) [reply tests/test_compiler.py]: I use `0.0.1` for testing the incompatible version and `0.4.0` for the compatible one. let me know if you think `999.0.0` is a better version :)

File: pennylane/compiler/compiler.py (hunk 7 above)
Comment (dime10) [reply pennylane/compiler/compiler.py:54]: It's initialized there, but this line was also resetting it every time. Is that not required?

Comment (maliasadi) [reply pennylane/compiler/compiler.py]: Oooooops! :((

**Review by dime10 — APPROVED** [review]
Great 🙂

File: pennylane/compiler/compiler.py [hunk 12]
```diff
@@ -41,7 +46,34 @@ class AvailableCompilers:
...
+def _check_compiler_version(name):
+    """Check if the installed version of the given compiler is greater than
+    or equal to the required minimum version.
+    """
+    if AvailableCompilers.compiler_checked[name]:
+        return  # Used the cached value!
+
+    if name == "catalyst":
+        installed_catalyst_version = metadata.version("pennylane-catalyst")
+        print(installed_catalyst_version)
```
Comment (maliasadi) [inline pennylane/compiler/compiler.py:69]: :t-rex:
